Playground Feature: Multiple contracts per account for Flipfest issue 21

soetop opened the pull request:
## Description

This PR is for Flipfest issue 21.

- [x] frontend: project default loading with multiple contracts per account 
- [x] frontend: contracts can be added, edited, deployed and deleted 
- [x] backend: support the above interactions

Check out the [demo video](https://youtu.be/PEB1A5GRAbA)

## To-dos and known-issues
- [x] export project folder
- [x] remove comments starting with '//soe'
- [x] fix Cadence code check for transactions as contract [declarations are not detected](https://user-images.githubusercontent.com/81572453/140640705-79b27723-0a13-444b-b7cb-5bb78b3ae1b0.png) - [fixed now](https://github.com/soetop/flow-playground/commit/ce3efb7d3e6c6b59492cea1093f88f436bc35a71)
- [x] write test coverage
   - [initial coverage](https://github.com/onflow/flow-playground-api/pull/14/commits/974ff2a9f9d3bd1bba7525cdc3dfbe30af797104)
   - [more coverage](https://github.com/soetop/flow-playground-api/commit/ee108a56cedd24c944963717d98a59c47c619b71)
- [x] code [documentation](https://laser-teal-a79.notion.site/Notes-for-changes-Flipfest-21-b2c49263e87a4551b57826070fe566a6)

## Submission Links & Documents

_Backend_: flow-playground-api has been forked and work is done on [flip-21 branch](https://github.com/soetop/flow-playground-api/tree/flip-21)

Run datastore emulator first
```
  gcloud beta emulators datastore start
```

Then run the server
```
  make run-datastore
```


_Frontend_: flow-playground has been forked and work is done on [flip-21 branch](https://github.com/soetop/flow-playground/tree/flip-21)

```
  npm install
  npm run start
```

soetop committed:
first functional commit v0.1

an early functional commit with a few loose ends

soetop committed:
Update projectDefault.ts

default project is updated to reflect multiple contracts per accounts to greet in various languages.

soetop committed:
code refactor for deploycontract

code refactor for deploycontract

soetop committed:
fix - set first contract of an account active

fix - set first contract of an account active if contractId is not defined

soetop committed:
minor code cleanup for routing

minor code cleanup for routing

soetop committed:
update queries to get deployedScript from contracts

update queries to get deployedScript from contracts

soetop committed:
export project with multiple contracts per account

export project with multiple contracts per account

soetop committed:
tweak contract tabs UI

tweak contract tabs UI to hint separation of layer and more tab-like with subtle shadow

soetop committed:
remove accountId from Contract struct

remove accountId from Contract struct - as it is not needed

index within Contract struct keeps track of a contract's account's index

MaxStalker commented on src/containers/Editor/components.tsx:
```diff
@@ -196,14 +213,52 @@ type EditorTitleProps = {
+            active.type == EntityType.Account ? active.index : null
+          }
+          onSelect={(_, id) => {
+            navigate(`/${projectPath}?type=account&id=${project.accounts[active.index].id}&contractId=${id}`)
```
We can probably move definition of `project.accounts[active.index].id` somewhere above for better code readability.
That field is being accessed at least 4 times in this `return` block

MaxStalker commented on src/layout/HeadingContainer.tsx:
```diff
@@ -0,0 +1,8 @@
+  margin: .2rem;
+  margin-top: 0rem;
+  border-bottom: 2px solid #f0f0f0;
+`;
```
Please,  add new line at the end of tile :)

soetop replied:
do you mean adding space beneath the title?

soetop replied:
i guess, you mean a new line in the .tsx file at the end, right?

soetop committed:
Update src/components/MenuList.tsx

Co-authored-by: Maksim Daunarovich <max.daunarovich@gmail.com>

soetop commented on src/components/Sidebar.tsx:
```diff
@@ -2,6 +2,7 @@ import React from "react";
 import { EntityType } from "providers/Project";
 import AccountList from "components/AccountList";
 import MenuList from "components/MenuList";
+//import ContractList from "components/ContractList";
```
not needed, i was initially putting contract menu items in the sidebar, instead of tabs

soetop commented on src/containers/Editor/components.tsx:
```diff
@@ -113,9 +118,15 @@ type EditorContainerProps = {
 
 function getActiveCode(project: Project, active: ActiveEditor): string {
+  //soe logging type,index,contractIndex inside getActiveCode
+  console.log(`type,index,contractIndex: ${active.type},${active.index},${active.contractIndex}`);
```
no, i will remove comments prefixed with //soe

soetop commented:
Hi @MaxStalker,

Thanks for code review. 

I have written a doc here: https://www.notion.so/Notes-for-changes-Flipfest-21-b2c49263e87a4551b57826070fe566a6?showMoveTo=true

I will resolve the changes you suggested. As well as remove my comments ( I needed them for my own reference otherwise I forget).

I have one question regarding Cadence code check on transactions as can be seen below. Deployed contracts are not found in declarations. However, transactions can be executed successfully as contracts actually exist.

![](https://user-images.githubusercontent.com/81572453/140642534-8932e885-1b48-46d8-ae91-bfc868230ed9.png)

Your guidance on this is much appreciated.

soetop committed:
regenerate graphql.tsx

regenerated graphql.tsx via `npx graphql-codegen --config codegen.yml`

soetop committed:
Merge branch 'flip-21' of https://github.com/soetop/flow-playground into flip-21

soetop committed:
pass the account's contracts code as one combined string

pass the account's contracts code as one combined string to Cadence server for check

this fix the frontend warning for missing declarations

soetop commented:
> Hi @MaxStalker,
> 
> Thanks for code review.
> 
> I have written a doc here: https://www.notion.so/Notes-for-changes-Flipfest-21-b2c49263e87a4551b57826070fe566a6?showMoveTo=true
> 
> I will resolve the changes you suggested. As well as remove my comments ( I needed them for my own reference otherwise I forget).
> 
> I have one question regarding Cadence code check on transactions as can be seen below. Deployed contracts are not found in declarations. However, transactions can be executed successfully as contracts actually exist.
> 
> ![](https://user-images.githubusercontent.com/81572453/140642534-8932e885-1b48-46d8-ae91-bfc868230ed9.png)
> 
> Your guidance on this is much appreciated.

This is fixed now with, this commit: https://github.com/soetop/flow-playground/commit/ce3efb7d3e6c6b59492cea1093f88f436bc35a71

Frontend Cadence code check refers to draft version of contract script/code. Hence, this fix combines an account's all draft contracts and return as one string.

soetop committed:
removed comments prefixed as '//soe'

removed comments prefixed as '//soe'

soetop committed:
making changes as suggested by Max

making changes as suggested by Max

https://github.com/onflow/flow-playground/pull/177#pullrequestreview-802863455

soetop committed:
removed 'draftCode' and 'deployedCode' from Account struct + code cleanup

removed 'draftCode' and 'deployedCode' from Account struct + code cleanup

soetop committed:
added more transactions in the default project

added more transactions in the default project - to highlight multiple contracts usage

soetop commented:
hello @MaxStalker and code owners,

I have made a few more commits. I would appreciate your reviews and will made necessary changes quickly.

For your reference, I have made documentation here:
https://laser-teal-a79.notion.site/Notes-for-changes-Flipfest-21-b2c49263e87a4551b57826070fe566a6

For quick reference, I have made a screen recording:
https://youtu.be/Wji5UD6HXZE

If you need quick response, I am available on Discord: soe#4875

soetop committed:
remove logging output

remove logging output

soetop committed:
improved contract deployment output to show contract name as well

improved contract deployment output to show contract name as well

soetop committed:
fix contract deployment issue upon first save of the project

fix contract deployment issue upon first save of the project

soetop committed:
tweak display of contracts in AccountList on the sidebar

tweak display of contracts in AccountList on the sidebar, display count of deployed contracts

soetop committed:
refine setting of contractIndex to 0 if it is invalid

refine setting of contractIndex to 0 if it is invalid

MaxStalker commented on src/components/MenuList.tsx:
```diff
@@ -71,12 +71,29 @@ const MenuList: React.FC<MenuListProps> = ({
 
-  const isScript = title.toLowerCase().includes("script");
-  const itemType = isScript ? EntityType.ScriptTemplate : EntityType.TransactionTemplate
+  //const isScript = title.toLowerCase().includes("script");
+  //const itemType = isScript ? EntityType.ScriptTemplate : EntityType.TransactionTemplate
```
Safe to delete? :)

soetop replied:
yes, you are right

soetop replied:
done

MaxStalker commented on src/components/AccountList.tsx:
```diff
@@ -17,7 +17,7 @@ function getDeployedContracts(account: Account): string {
     contract => contract.split(".").slice(-1)[0]
   );
-  return contracts.join(", ");
+  return "("+ contracts.length +") contracts ... "+ contracts.join(", ");
```
Maybe we can render this as a list instead of simply a string, cause it will be cutout most of the time 🤔

soetop replied:
how would you want to display on the UI?

MaxStalker replied:
It's better to display which contracts are currently deployed there than simply a number :)
Maybe `<ul><li/></ul>` block?
Or simply make `<p>` block with `<a>` links, which would open tab for that contract?

soetop replied:
Made the changes in this commit: https://github.com/onflow/flow-playground/pull/177/commits/1df5ce5050fc693c3f1fc0a9ef7109d3e25c32f2
But it looks a bit weird when there are more than 2 deployed contracts.

<img width="228" alt="Screen Shot 2021-11-21 at 7 09 55 PM" src="https://user-images.githubusercontent.com/81572453/142762182-66b7642d-a2c1-4292-8a7c-6feaffcc5c59.png">

MaxStalker commented on src/containers/Editor/components.tsx:
```diff
@@ -170,7 +179,17 @@ const EditorContainer: React.FC<EditorContainerProps> = ({
       }
-      return project.accounts[index].draftCode
+
+      // combine the account's contracts as one string
+      // and return for cadence server check
+      var combinedCode = '';
+      for (let i = 0; i < project.contracts.length; i++) {
+        if(project.contracts[i].index == index) {
+          combinedCode += project.contracts[i].script;
+        }
+      }
+
+      return combinedCode;
```
Why we are doing this? 🤔

soetop replied:
code is passed to Cadence server to check for available declarations/contracts deployed at an account. this is to fix the earlier issue noted here: https://user-images.githubusercontent.com/81572453/140640705-79b27723-0a13-444b-b7cb-5bb78b3ae1b0.png

soetop replied:
previous only the account's code (1 contract) was passed for declaration. since we now have multiple contracts (hence multiple declarations) we need to pass all contracts of an account, combined. this is just for frontend check.

MaxStalker commented on src/providers/Project/projectHooks.ts:
```diff
@@ -22,7 +22,13 @@ function cloneProject(client: any, project: Project) {
+    project.accounts.map(acc => acc.state),
+
+    project.contracts.map(con => ({
+      code: con.script,
```
Back to that `code` vs `script` name :)

soetop replied:
yes, i will refactor that

soetop replied:
this commit addressed this: [300f2c6ce469f16f009cc8eda72d105a7fad8f8f](https://github.com/soetop/flow-playground/commit/300f2c6ce469f16f009cc8eda72d105a7fad8f8f)

MaxStalker commented on src/util/generator.ts:
```diff
@@ -315,13 +315,15 @@ export const createZip = async (
+  for (let i = 0; i < contracts.length; i++) {
+    const contract = contracts[i];
+    const name = getContractName(contract.script);
+    const fileName = `cadence/contracts/0x0${contract.index+1}/${name}.cdc`;
```
I would probably make this one a bit more elaborate and only add prefix/suffix if there is name collision, but this will do for now :)

soetop replied:
yes, noted. a follow up question, is it possible to import the exported project folder somewhere?

soetop commented on src/api/apollo/mutations.ts:
```diff
@@ -48,41 +48,68 @@ export const SET_ACTIVE_PROJECT = gql`
+  mutation CreateContract(
+    $projectId: UUID!
+    $index: Int!
+    $script: String!
```
noted, i will refactor that

soetop replied:
do you want me to refactor `index` to `accountIndex` as well?

soetop replied:
this commit addressed this: [300f2c6ce469f16f009cc8eda72d105a7fad8f8f](https://github.com/soetop/flow-playground/commit/300f2c6ce469f16f009cc8eda72d105a7fad8f8f)

soetop committed:
refactor Contract struct naming

refactor Contract struct naming

`Index` to `AccountIndex`
`Script` to `Code`
`DeployedScript` to `DeployedCode`

soetop committed:
added new line at EOF

added new line at EOF

soetop committed:
removed 2 commented lines

removed 2 commented lines

soetop committed:
removed export/embed button for account menu item

removed export/embed button for account menu item

soetop committed:
remove unused variable

remove unused variable

MaxStalker commented on src/providers/Project/projectHooks.ts:
```diff
@@ -22,7 +22,13 @@ function cloneProject(client: any, project: Project) {
-    project.accounts.map(acc => acc.draftCode),
+    project.accounts.map(acc => acc.state),
+
+    project.contracts.map(con => ({
+      code: con.code,
+      title: con.title,
+      accountIndex: con.accountIndex
+    })),
```
```suggestion
    project.contracts.map(({ code, title, accountIndex }) => ({ code, title, accountIndex })),
```

MaxStalker replied:
Maybe alter scripts and transaction templates in the same manner? 🤔

soetop replied:
yes

soetop replied:
this commit fixed this: https://github.com/onflow/flow-playground/pull/177/commits/b7acfcc9e8149daad022f7bb47aac1296f7a2eca

MaxStalker commented on src/components/AccountList.tsx:
```diff
@@ -65,8 +63,6 @@ const AccountList: React.FC = () => {
                   <small>{contractName || '--'}</small>
                 </Stack>
-
-                {isActive && <ExportButton id={account.id} typeName={typeName}/>}
```
Do we have this ExportButton in other place?

soetop replied:
yes, adding in ContractList.tsx for active contract tab item, tweaking styling now

soetop replied:
this commit fixed this: https://github.com/onflow/flow-playground/pull/177/commits/3720f567eb86b83422308c80a1aa30903e2249dc

soetop committed:
update mapping with shorthand as suggested by Max

update mapping with shorthand as suggested by Max

soetop committed:
add export/embed button to ContractList tab items

add export/embed button to ContractList tab items

soetop committed:
redirect for type=contract

redirect for type=contract, this is referenced from embed code

soetop committed:
update display of deployed contracts as list

update display of deployed contracts as list

soetop committed:
update styling for AccountList list item

update styling for AccountList list item

soetop committed:
tweak AccountList active item

to show active if active.index correspond to the account as well

soetop committed:
remove logging in Project/index.tsx

remove logging in Project/index.tsx

MaxStalker reviewed (approved):
Looks good :)
Thank you for bringing those improvements @soetop ! 🙇
